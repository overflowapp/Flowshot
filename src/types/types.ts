export interface ClickData {
    click: boolean;
    payload: {
        pageX: number;
        pageY: number;
        screenX: number;
        screenY: number;
        clientX: number;
        clientY: number;
        boundingRect: {
            x: number;
            y: number;
            h: number;
            w: number;
        };
    };
}

export interface TabShot {
    title: string;
    dataURI: string;
    size: Size;
}

export interface Shot {
    date: number;
    tab: TabShot;
    event: ClickData;
}

export interface Size {
    h: number;
    w: number;
}

export interface Session {
    date: number;
    shots: Shot[];
}

export enum ChromeTabStatus {
    'loading' = 'loading',
    'complete' = 'complete',
}

export enum SessionState {
    'stopped' = 0,
    'started' = 1,
    'discarded' = 2,
}
