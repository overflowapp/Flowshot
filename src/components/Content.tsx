import * as React from 'react';

interface AppProps {}

interface AppState {
    recording: boolean;
}

export default class Content extends React.Component<AppProps, AppState> {
    state = {
        recording: false,
    };

    render() {
        return <p>Flowshot popup</p>;
    }
}
