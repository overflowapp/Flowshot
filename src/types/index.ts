import EventBus from '../EventBus';

export interface Screenshot {
    title: string;
    date: number;
    dataURI: string;
    bounds: {
        h: number;
        w: number;
    };
}

declare global {
    interface Window {
        FlowshotEvents: EventBus;
    }
}
