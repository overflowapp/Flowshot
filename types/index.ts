import EventBus from '../src/EventBus';

export interface Session {
    date: number;
    data: SessionData[];
}

export interface SessionData {
    date: number;
    screen: {
        tab: string;
        dataURI: string;
        dimensions: {
            w: number;
            h: number;
        };
    };
    click: {
        pageX: number;
        pageY: number;
        screenX: number;
        screenY: number;
        clientX: number;
        clientY: number;
        boundingRect: {
            x: number;
            y: number;
            h: number;
            w: number;
        };
    };
}

export interface Recording {
    status: RecordingStatus;
    events: RecordingEvent[];
}

export interface RecordingEvent {
    tab: string;
    image: string;
    bounds: {
        pageX: number;
        pageY: number;
        clientX: number;
        clientY: number;
    };
}

export enum RecordingStatus {
    'stopped' = 0,
    'started' = 1,
    'discarded' = 2,
}

export enum ChromeTabStatus {
    'loading' = 'loading',
    'complete' = 'complete',
}

declare global {
    interface Window {
        FlowshotEvents: EventBus;
    }
}
