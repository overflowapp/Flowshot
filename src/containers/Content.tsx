import * as React from 'react';
import { Screenshot } from '../types';
import TabHelper from '../utils/TabHelper';
import { SessionStarted, SessionDiscarded, SessionStopped } from '../components';
import { SessionState, RequestType } from '../types/types';

interface AppProps {}

interface AppState {
    status: SessionState;
    screenshots: Screenshot[];
}

export default class Content extends React.Component<AppProps, AppState> {
    state = {
        status: SessionState.stopped,
        screenshots: [],
    };

    componentWillMount() {
        this.listen();
    }

    componentDidMount() {
        chrome.runtime.sendMessage({ popupMounted: true });
        this.setFromLocalState();
    }

    setFromLocalState = () => {
        chrome.storage.local.get(['recordingState'], result => {
            result.recordingState &&
                this.setState({
                    status: result.recordingState as SessionState,
                });
        });
    };

    listen() {
        chrome.runtime.onMessage.addListener((request, sender, respond) => {
            switch (true) {
                case request.type === RequestType.NEW_IMAGE:
                    this.setState(prevState => ({
                        screenshots: [
                            ...prevState.screenshots,
                            {
                                title: 'any',
                                date: Date.now(),
                                dataURI: request.payload.dataURI,
                                bounds: { ...request.payload.dimensions },
                            },
                        ],
                    }));
                    break;
            }
        });
    }

    setRecordingState = (status: SessionState) => {
        TabHelper.getCurrentTab().then(tab => {
            chrome.tabs.sendMessage(tab.id, {
                changeRecordingState: true,
                recordingState: status === SessionState.started ? SessionState.started : SessionState.stopped,
            });

            chrome.storage.local.set({ recordingState: status });
            chrome.runtime.sendMessage({ recordingState: status });

            this.setState({
                status,
            });
        });
    };

    PopupContent = () => {
        switch (this.state.status) {
            case SessionState.started:
                return <SessionStarted setRecordingState={this.setRecordingState} />;
            case SessionState.stopped:
                return <SessionStopped setRecordingState={this.setRecordingState} />;
            case SessionState.discarded:
                return <SessionDiscarded setRecordingState={this.setRecordingState} />;
        }
    };

    render() {
        const { screenshots } = this.state;
        const SessionState = this.PopupContent;

        return (
            <React.Fragment>
                <SessionState />
                {screenshots.map(s => (
                    <img key={s.date} height={140} src={s.dataURI} />
                ))}
            </React.Fragment>
        );
    }
}
