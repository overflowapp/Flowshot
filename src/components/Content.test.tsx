import * as React from 'react';
import { shallow, mount } from 'enzyme';
import Content from './Content';
import { SessionStatus } from '../../types';

describe('Content', function() {
    it('should have a stopped session status', function() {
        expect(shallow(<Content />).state('status')).toBe(SessionStatus.stopped);
    });

    it('should have no screenshots', function() {
        expect((shallow(<Content />).state('screenshots') as []).length).toBe(0);
    });

    it('should have a paragraph', function() {
        expect(mount(<Content />).find('p').length).toBe(1);
    });

    it('pass snapshot', function() {
        const tree = shallow(<Content />);
        expect(tree).toMatchSnapshot();
    });
});
