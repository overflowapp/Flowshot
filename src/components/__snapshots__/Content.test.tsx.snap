// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Content pass snapshot 1`] = `
<p>
  Flowshot popup
</p>
`;
